Type examples/

frrad opened the pull request:
Adds typing to files in `examples/`.

Also removed several more `pulp/` files from the mypy blacklist, adding `type: ignore`s where necessary.

pchtsp commented on examples/furniture.py:
```diff
@@ -14,14 +14,14 @@
     [3, 1.5],  # Polisher
 ]
-activity = makeDict([Resources, Chairs], activity)
+activity_dict = makeDict([Resources, Chairs], activity)
```
I'm curious, is this some kind of naming convention to recognize the type objects? Why rename all dictionaries to have a `_dict` at the end?

frrad replied:
I changed the name of this variable because previously this line was changing the type of `activity` from a list to a dict. The `_dict` is a bit lazy admittedly. Very open to better name suggestions :)

pchtsp commented on examples/SpongeRollProblem4.py:
```diff
@@ -80,14 +82,14 @@ class Pattern:
     lenOpts = [5, 7, 9]
 
-    def __init__(self, name, lengths=None):
+    def __init__(self, name: str, lengths: List[int]) -> None:
```
shouldn't we keep the `=None` here?

frrad replied:
I removed it because the `zip` below on L87 fails to typecheck if `lengths` is `None`. Since this code is presumably working I took that to mean that this is never used that way in practice.

pchtsp commented on examples/CGcolumnwise.py:
```diff
@@ -19,21 +22,21 @@ class Pattern:
     lenOpts = ["5", "7", "9"]
     numPatterns = 0
 
-    def __init__(self, name, lengths=None):
```
also here, should we keep the `=None`?

frrad replied:
as above

pchtsp commented on examples/BeerDistributionProblem_resolve.py:
```diff
@@ -68,10 +68,10 @@
 # The problem data is written to an .lp file
 prob.writeLP("BeerDistributionProblem.lp")
-for demand in range(500, 601, 10):
+for demand_value in range(500, 601, 10):
```
same as with `_dict`, is this some kind of naming convention based on the type of the object?

frrad replied:
Similar to previously there is already a `demand` in scope and this was me fixing the collision.

pchtsp reviewed (commented):
Thanks! I added minor questions about the renaming of some objects. I'm curious what tool renames objects based on their implied type. It sounds a bit weird.

christiansegercrantz commented:
> Thanks! I added minor questions about the renaming of some objects. I'm curious what tool renames objects based on their implied type. It sounds a bit weird.

This naming convention is in my opinion a bit outdated with modern typing. Instead of changing names to indicate type it would be better, as is being done in the linked PR, to indicate the type properly with typing.

MBradbury commented:
Few comments based on #807:
1. Why use `List` and not `list` now that 3.8 has been dropped?
2. I tried to check the type hints from the previous PR and different tools work differently. For example, I use pyright which is typically significant stricter than mypy and pyright still identifies some issues.
3. In #807, I also did a lot of typing work in examples/, its not likely that I will get round to fixing that PR, so might be worth taking a look and extracting what is needed.

MBradbury commented on pulp/pulp.py:
```diff
@@ -144,28 +144,28 @@
 try:  # allow Python 2/3 compatibility
     maketrans = str.maketrans
 except AttributeError:
-    from string import maketrans
```
Why not just change to be: `from string import maketrans`, Python 2 is not supported

MBradbury commented on pulp/pulp.py:
```diff
@@ -144,28 +144,28 @@
 
 _DICT_TYPE = dict
 
 if sys.platform not in ["cli"]:
```
There is no `sys.platform` which is `cli` any more. I presume this was IronPython. Why not simplify and just remove `_DICT_TYPE` as per https://github.com/coin-or/pulp/pull/807/files#diff-b2a258edfe0fa294f032a978656a4082842558faa98b0950d9f68c50f546bbe2L151

MBradbury commented on pulp/pulp.py:
```diff
@@ -1063,7 +1063,7 @@ def __init__(self, e=None, sense=const.LpConstraintEQ, name=None, rhs=None):
         """
         self.expr = e if isinstance(e, LpAffineExpression) else LpAffineExpression(e)
         self.name = name
-        self.constant: float = self.expr.constant
```
This just hides a potential bug. Better to fix the annotations

https://github.com/coin-or/pulp/pull/807/files#diff-b2a258edfe0fa294f032a978656a4082842558faa98b0950d9f68c50f546bbe2R169

https://github.com/coin-or/pulp/pull/807/files#diff-b2a258edfe0fa294f032a978656a4082842558faa98b0950d9f68c50f546bbe2R1212

MBradbury replied:
In my PR, I added some type aliases, which I then used where appropriate.
```python
LptNumber: TypeAlias = int | float
LptItem: TypeAlias = Union[LptNumber, "LpVariable"]
LptExpr: TypeAlias = Union[LptItem, "LpAffineExpression"]
LptConstExpr: TypeAlias = Union[LptNumber, "LpAffineExpression"]
LptExprConstraint: TypeAlias = Union[LptExpr, "LpConstraint"]
LptVars: TypeAlias = "LpVariable"
```

MBradbury commented on pulp/pulp.py:
```diff
@@ -1218,7 +1218,7 @@ def __truediv__(self, other):
     def valid(self, eps: float = 0) -> bool:
         val = self.value()
         if self.sense == const.LpConstraintEQ:
-            return abs(val) <= eps
```
Again, just hides the bug: https://github.com/coin-or/pulp/pull/807/files#diff-b2a258edfe0fa294f032a978656a4082842558faa98b0950d9f68c50f546bbe2R1370

MBradbury replied:
My links don't seem to have worked, there needs to be a check like:

```python
val = self.value()
if val is None:
    return False
```

MBradbury commented:
There are some easy wins in the type hinting space, but as I found in #807, there are some architectural issues with LpElement, LpVariable and LpConstraintVar that need to be fixed. Also FractionElasticSubProblem just seems fundamentally broken.

Then there is also the question of what type the coefficient should be? int, float, decimal, numpy int/float types, or what combination of these?

MBradbury commented on examples/SpongeRollProblem5.py:
```diff
@@ -21,23 +21,25 @@
 while morePatterns == True:
     # Solve the problem as a Relaxed LP
     duals = masterSolve(Patterns, rollData)
+    assert isinstance(duals, dict)
```
Surely, the return type is a tuple as per `masterSolve`?

frrad commented on examples/SpongeRollProblem5.py:
```diff
@@ -21,23 +21,25 @@
     Patterns += [Pattern("P" + str(len(Patterns)), i)]
 
 # This loop will be repeated until morePatterns is set to False
 while morePatterns == True:
```
I am trying to keep the scope of this PR as limited as possible. There's a ton of less-than-ideal code here. If I started fixing things I don't know where I would be able to stop.

MBradbury replied:
Same problem I had in #807 :)

frrad commented:
👋 thanks for all the reviews.

To clear up some possible confusion - this PR is intended to be the minimum possible change to add types to the `examples/` directory. It is explicitly _not_ trying to do things like fix types in `pulp/` or improve the non-type code in `examples`. I think structuring things this way will lead to a change which is easy to review and a pareto improvement over the current situation. There will still definitely remain a lot of room to improve things afterwards, but by trying to be focused here I think we can hopefully get this one in without too much back and forth and make the next PR easier for whoever ends up making it.


Additional responses inline: 


> > Thanks! I added minor questions about the renaming of some objects. I'm curious what tool renames objects based on their implied type. It sounds a bit weird.
> 
> This naming convention is in my opinion a bit outdated with modern typing. Instead of changing names to indicate type it would be better, as is being done in the linked PR, to indicate the type properly with typing.

Sorry for the confusion caused by my clunky variable names. They are not meant to convey the type, just fix the variable type change issues. These variables all actually typecheck fine without adding explicit hints.

> re: changes to `pulp`

These are all great questions. The intention of this PR is not to fix any type problems in `pulp/`. I just added whatever `type: ignore` were necessary to allow me to remove these files from the mypy blacklist. I definitely agree that a lot more work is needed to actually fix the issues there.

MBradbury commented:
> > re: changes to `pulp`
> 
> These are all great questions. The intention of this PR is not to fix any type problems in `pulp/`. I just added whatever `type: ignore` were necessary to allow me to remove these files from the mypy blacklist. I definitely agree that a lot more work is needed to actually fix the issues there.

Would it be better to **not** add the `# type: ignore`s then? By adding them potential bugs are covered up. IMO, better to let the tests fail and then fix the bug than to hide the issue.

frrad commented:
In general unit tests / CI work best if you keep them in a state where you always expect them to pass. You want their binary pass / fail state to be a useful signal of whether or not a new change introduces new bugs. If you add tests which you expect to fail it can confuse things quite a bit.

The type ignore statements are a useful crutch to be able to make incremental improvements without having to fix everything at once. With their help we can do 

1. The current state mypy doesn't even consider these files and CI gives us no information about what changes do to typing in them
2. After we add ignores readers of the code will at least be aware of existing type issues. 
3. Any future PRs which touch these files will fail CI if they introduce new type errors.
4. New PRs can be made which fix the existing type issues and remove the ignore comments. Reviewers will be able to be confident in the correctness of these PRs because CI status will tell them whether or not mypy passes

christiansegercrantz commented:
> > > Thanks! I added minor questions about the renaming of some objects. I'm curious what tool renames objects based on their implied type. It sounds a bit weird.
> > 
> > 
> > This naming convention is in my opinion a bit outdated with modern typing. Instead of changing names to indicate type it would be better, as is being done in the linked PR, to indicate the type properly with typing.
> 
> Sorry for the confusion caused by my clunky variable names. They are not meant to convey the type, just fix the variable type change issues. These variables all actually typecheck fine without adding explicit hints.

Makes sense, didn't read the code in larger context. Good change then, at some point those should probably have completely different names to convey their meaning better. But a good first step.

frrad commented:
merged master to pick up https://github.com/coin-or/pulp/pull/820 - looks like I have some work to do

frrad commented:
That wasn't too bad at least 😅 

@pchtsp I think I have addressed all your comments. Please let me know if there is anything more I can answer or change.

There are some references to these files in the documentation by line number like this: https://github.com/coin-or/pulp/blob/f626d34f9b345ffc8bf405f9711d160fe20a4f1d/doc/source/CaseStudies/a_blending_problem.rst?plain=1#L97-L98

which may need to be updated if things have shifted around. I'm happy to have a look at this as well as part of this PR. I want to wait until the actual code changes look good first though in case the lines end up moving around more.

pchtsp commented:
thanks @frrad . I don't mind if we do not have a perfect state now. Anything that gets as closer is good. 
@christiansegercrantz , @MBradbury thanks for your help, if you can make any improvement on this on a separate PR (e.g., through https://github.com/coin-or/pulp/pull/807/) I can review it and merge it. Again, if we do it in pieces, we will probably finish sooner.

Thanks again
